letsencrypt/client/tests/reverter_test.py

```diff
@@ -0,0 +1,451 @@
+
+    def test_recover_checkpoint_rm_failure(self):
+        self.reverter.add_to_temp_checkpoint(self.sets[0], "temp save")
+
+        with mock.patch("letsencrypt.client.reverter.shutil."
+                        "rmtree") as mock_rmtree:
+            mock_rmtree.side_effect = OSError("Cannot remove tree")
+            self.assertRaises(errors.LetsEncryptReverterError,
+                              self.reverter.revert_temporary_config)
+
+    @mock.patch("letsencrypt.client.reverter.logging.warning")
+    def test_recover_checkpoint_missing_new_files(self, mock_warn):
+        self.reverter.register_file_creation(
+            True, os.path.join(self.dir1, "missing_file.txt"))
+        self.reverter.revert_temporary_config()
+        self.assertEqual(mock_warn.call_count, 1)
+
+    @mock.patch("letsencrypt.client.reverter.os.remove")
+    def test_recover_checkpoint_remove_failure(self, mock_remove):
+        self.reverter.register_file_creation(True, self.config1)
+        mock_remove.side_effect = OSError("Can't remove")
+        self.assertRaises(errors.LetsEncryptReverterError,
+                          self.reverter.revert_temporary_config)
+
+    def test_recovery_routine_temp_and_perm(self):
+        # Register a new perm checkpoint file
+        config3 = os.path.join(self.dir1, "config3.txt")
+        self.reverter.register_file_creation(False, config3)
+        update_file(config3, "This is a new perm file!")
+
+        # Add changes to perm checkpoint
+        self.reverter.add_to_checkpoint(self.sets[0], "perm save1")
+        update_file(self.config1, "updated perm config1")
+        self.reverter.add_to_checkpoint(self.sets[1], "perm save2")
+        update_file(self.config2, "updated perm config2")
+
+        # Add changes to a temporary checkpoint
+        self.reverter.add_to_temp_checkpoint(self.sets[0], "temp save1")
+        update_file(self.config1, "second update now temp config1")
+
+        # Register a new temp checkpoint file
+        config4 = os.path.join(self.dir2, "config4.txt")
+        self.reverter.register_file_creation(True, config4)
+        update_file(config4, "New temporary file!")
+
+        # Now erase everything
+        self.reverter.recovery_routine()
+
+        # Now Run tests
+        # These were new files.. they should be removed
+        self.assertFalse(os.path.isfile(config3))
+        self.assertFalse(os.path.isfile(config4))
+
+        # Check to make sure everything got rolled back appropriately
+        self.assertEqual(read_in(self.config1), "directive-dir1")
+        self.assertEqual(read_in(self.config2), "directive-dir2")
+
+
+class TestFullCheckpointsReverter(unittest.TestCase):
+    # pylint: disable=too-many-instance-attributes
+    """Tests functions having to deal with full checkpoints."""
+    def setUp(self):
+        from letsencrypt.client.reverter import Reverter
+        # Disable spurious errors...
+        logging.disable(logging.CRITICAL)
```

Comment:
Same as in https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23518302